pydantic/dataclasses.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass as dc
+from .main import validate_model, create_model
+
+
+def pydantic_model(cls):
+    fields = {name: (field.type, field.default)
+              for name, field
+              in cls.__dataclass_fields__.items()}
+    model = create_model('', **fields)
+    return model
+
+
+def dataclass(_cls=None, *, init=True, repr=True, eq=True,
+              order=False, unsafe_hash=False, frozen=False):
+    def wrap(cls):
+        post_init_orig = getattr(cls, '__post_init__', lambda self: None)
+        cls.__post_init__ = post_init_orig
+        cls = dc(init=init, repr=repr, eq=eq, order=order,
+                 unsafe_hash=unsafe_hash, frozen=frozen)(cls)
+        model = pydantic_model(cls)
```

Comment:
we have a strange mixture here of relying on closure and not. Maybe it would be clearer if we did

```py
cls.__model_cls__ = pydantic_model(cls)
```
Then could move `set_attr` and `post_init` definitions outside this function to the global scope.